Composer/packages/client/src/recoilModel/dispatchers/auth.ts

```diff
@@ -0,0 +1,156 @@
+/* eslint-disable react-hooks/rules-of-hooks */
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import formatMessage from 'format-message';
+import { CallbackInterface, useRecoilCallback } from 'recoil';
+import { AzureTenant } from '@botframework-composer/types';
+
+import {
+  userShouldProvideTokens,
+  isShowAuthDialog,
+  getTokenFromCache,
+  setTenantId,
+  getTenantIdFromCache,
+} from '../../utils/auth';
+import {
+  showAuthDialogState,
+  primaryTokenState,
+  graphTokenState,
+  availableTenantsState,
+  showTenantDialogState,
+  currentTenantState,
+} from '../atoms/appState';
+import { AuthClient } from '../../utils/authClient';
+import storage from '../../utils/storage';
+
+import { addNotificationInternal, createNotification } from './notification';
+
+export const authDispatcher = () => {
+  const setShowAuthDialog = useRecoilCallback((callbackHelpers: CallbackInterface) => (show: boolean) => {
+    callbackHelpers.set(showAuthDialogState, show);
+  });
+
+  const setPrimaryToken = useRecoilCallback((callbackHelpers: CallbackInterface) => (token: string) => {
+    callbackHelpers.set(primaryTokenState, token);
+    storage.set('accessToken', token);
+  });
+
+  const setGraphToken = useRecoilCallback((callbackHelpers: CallbackInterface) => (token: string) => {
+    callbackHelpers.set(graphTokenState, token);
+    storage.set('graphToken', token);
+  });
+
+  const setCurrentTenant = useRecoilCallback(
+    (callbackHelpers: CallbackInterface) => async (tenant: string, notify = true) => {
+      callbackHelpers.set(currentTenantState, tenant);
+      setTenantId(tenant);
+      if (tenant) {
+        // get arm token for tenant
+        try {
+          const token = await AuthClient.getARMTokenForTenant(tenant);
+          if (token) {
```

Comment:
If the token is undefined or empty do you want to set the primary token to empty anyway (vs. leaving the existing value in there)?

Reply:
good call.  getArmTokenForTenant will throw if it fails, so the catch -- which contains a call to reset the token -- should fire.  But I added an else just in case to throw if token is undefined and make sure!